Animation

johnhaddon opened the pull request:
This contains baby steps towards adding keyframe animation to Gaffer :

- Context now has setTime()/getTime() and setFramesPerSecond()/getFramesPerSecond() whereas before it only had setFrame()/getFrame(). The framesPerSecond field is used to map between time and frame values - currently with time==0 at frame==0. I have a feeling some packages go for time==0 at frame==1, but I think I prefer the simplicity of this (and the backwards compatibility with the SceneReader and AlembicReader behaviour) - does anyone want to make a case for the opposite?
- New Animation node type provides the ability to store animation curves and apply them to plugs, and a basic API for adding and removing keys.
- Right click plug popup menu in UI allows keys to be added and removed, and to jump backwards and forwards to the previous and next keys.
- NumericPlugValueWidget and BoolPlugValueWidget automatically add keys when they are edited on a plug with an existing animation.

There's plenty still to do :

- Curve editor.
- Cubic interpolation. I don't think this would take too long, but without a curve editor it's a little hard to visualise.
- UI for keying x/y/z and r/g/b all at once (it's just limited to one plug at a time right now).
- Removing the last keyframe via the UI should probably also remove the animation entirely. I think maybe we need a higher level AnimationAlgo API for this, since it feels odd that a CurvePlug would delete itself while you're calling a method on it.

Let me know if you feel that any of that is necessary immediately or if these baby steps are useful enough on their own.

andrewkaufman commented:
I don't know what is going on exactly, but when I grab your branch using the command lines:

```
git checkout -b johnhaddon-animation master
git pull https://github.com/johnhaddon/gaffer.git animation
```

I end up with NumericWidget.py in a bad state, not as it shows in the diff on GitHub... I had to edit it locally to get something usable (which seems to work) but my entire __editingFinished function looks different, so I'm a bit scared...

```
diff --git a/python/GafferUI/NumericWidget.py b/python/GafferUI/NumericWidget.py
index 5253c98..a041f4e 100644
--- a/python/GafferUI/NumericWidget.py
+++ b/python/GafferUI/NumericWidget.py
@@ -251,7 +251,7 @@ class NumericWidget( GafferUI.TextWidget ) :
                # zeroes in order to achieve consistent editing. Revert
                # back to our standard form now so we don't leave it in
                # this state.
-               self.setText( self.__valueToText( self.getValue() ) )
+               self.setText( self.__valueToString( self.getValue() ) )
 
                self.__emitValueChanged( self.ValueChangedReason.Edit )
```

andrewkaufman commented:
Also, just playing with a keyframed translate.x on a Transform, I seem to have gotten into a state (after adding and removing several keys) where I can neither jumpToNext/Previous nor Remove, but the UI is still showing green like I have a key.

andrewkaufman commented:
Do you find it a bit odd that we can still set keys on Presets Only dropdowns like Sphere type or Transform space? I guess those are IntPlugs in the end, so it makes sense that we can, but seems a bit odd for the UI to make it so easy... maybe it doesn't matter though.

andrewkaufman commented:
Other than those, good to go. I'm not sure what to do about the crazy "I can't download an accurate representation of your branch" thing... is it rebased on the current master? Maybe the auto-merge screws up in a way it didn't consider a conflict? Or maybe it's user error and I just need to go home for the day...

andrewkaufman commented:
Oh, I forgot to verify before I left the office... can you double check that you're able to:

- set keys on a plug inside a box, export a reference, and check they stick
- promote that plug to the box, exposing the keys you set, add some more keys after promotion, export and check they all stick
- as a user of the Reference, see if you can add even more keys that serialize correctly

johnhaddon commented:
> I end up with NumericWidget.py in a bad state, not as it shows in the diff on GitHub... I had to edit it locally to get something usable (which seems to work) but my entire __editingFinished function looks different, so I'm a bit scared…

I've rebased and sorted this out.

> Also, just playing with a keyframed translate.x on a Transform, I seem to have gotten into a state (after adding and removing several keys) where I can neither jumpToNext/Previous nor Remove, but the UI is still showing green like I have a key.

This is most likely what I was talking about when I said "Removing the last keyframe via the UI should probably also remove the animation entirely. I think maybe we need a higher level AnimationAlgo API for this, since it feels odd that a CurvePlug would delete itself while you're calling a method on it.".

> Do you find it a bit odd that we can still set keys on Presets Only dropdowns like Sphere type or Transform space?

I can imagine wanting to keyframe a presets only thing, so I don't think it's too weird. What would be odd though is keyframing them with the default type as we're currently doing - it would only make sense to keyframe them with Step. I think I'm inclined to deal with this at a later date though - again perhaps in some higher level AnimationAlgo API.

>Oh, I forgot to verify before I left the office... can you double check that you're able to:

>- set keys on a plug inside a box, export a reference, and check they stick
- promote that plug to the box, exposing the keys you set, add some more keys after promotion, export and check they all stick
- as a user of the Reference, see if you can add even more keys that serialise correctly

I've added a test asserting that the first thing works, but unfortunately also asserting that the second two can never ever work. The first of the two would break the encapsulation that Boxes provide, and the second would require some kind of crazy merge-all-the-keys-oh-and-probably-support-deletion-of-keys-too reference edit system, whereas the whole point of the encapsulation is to avoid a need for that kind of thing. I'm presuming that since you've asked about it, we have a use case for this kind of thing?

andrewkaufman commented:
No use case in mind, it was mostly just randomly thinking about the sorts of things I can see users doing without thinking about why it may or may not be a good idea. I'm happy to wait for that to come up in a real situation then.

johnhaddon commented:
That's a bit of a relief! When an actual situation comes up perhaps we'll find that it can be addressed by thinking of things in slightly different terms...
